//------------------------------------------------------------------------------
// protobuf specification for Accelerator Ring Group related APIs
//------------------------------------------------------------------------------

syntax = "proto3";

import "types.proto";
//import public "gogo.proto";

package accelRGroup;
option go_package="halproto";

service AccelRGroup {
    rpc AccelRGroupAdd (AccelRGroupAddRequestMsg) returns (AccelRGroupAddResponseMsg) {}
    rpc AccelRGroupDel (AccelRGroupDelRequestMsg) returns (AccelRGroupDelResponseMsg) {}
    rpc AccelRGroupRingAdd (AccelRGroupRingAddRequestMsg) returns (AccelRGroupRingAddResponseMsg) {}
    rpc AccelRGroupRingDel (AccelRGroupRingDelRequestMsg) returns (AccelRGroupRingDelResponseMsg) {}
    rpc AccelRGroupResetSet (AccelRGroupResetSetRequestMsg) returns (AccelRGroupResetSetResponseMsg) {}
    rpc AccelRGroupEnableSet (AccelRGroupEnableSetRequestMsg) returns (AccelRGroupEnableSetResponseMsg) {}
    rpc AccelRGroupPndxSet (AccelRGroupPndxSetRequestMsg) returns (AccelRGroupPndxSetResponseMsg) {}
    rpc AccelRGroupInfoGet (AccelRGroupInfoGetRequestMsg) returns (AccelRGroupInfoGetResponseMsg) {}
    rpc AccelRGroupIndicesGet (AccelRGroupIndicesGetRequestMsg) returns (AccelRGroupIndicesGetResponseMsg) {}
    rpc AccelRGroupMetricsGet (AccelRGroupMetricsGetRequestMsg) returns (AccelRGroupMetricsGetResponseMsg) {}
    rpc AccelRGroupMiscGet (AccelRGroupMiscGetRequestMsg) returns (AccelRGroupMiscGetResponseMsg) {}
}

message AccelRGroupAddRequest {
    string rgroup_name = 1;
    fixed64 metrics_mem_addr = 2;
    uint32 metrics_mem_size = 3;
}

message AccelRGroupAddRequestMsg {
    repeated AccelRGroupAddRequest request = 1;
}

message AccelRGroupAddResponse {
    types.ApiStatus api_status = 1;
}

message AccelRGroupAddResponseMsg {
    repeated AccelRGroupAddResponse response = 1;
}

message AccelRGroupDelRequest {
    string rgroup_name = 1;
}

message AccelRGroupDelRequestMsg {
    repeated AccelRGroupDelRequest request = 1;
}

message AccelRGroupDelResponse {
    types.ApiStatus api_status = 1;
}

message AccelRGroupDelResponseMsg {
    repeated AccelRGroupDelResponse response = 1;
}

message AccelRGroupRingAddRequest {
    string rgroup_name = 1;
    string ring_name = 2;
    uint32 ring_handle = 3;
}

message AccelRGroupRingAddRequestMsg {
    repeated AccelRGroupRingAddRequest request = 1;
}

message AccelRGroupRingAddResponse {
    types.ApiStatus api_status = 1;
    uint32 ring_handle = 2;
}

message AccelRGroupRingAddResponseMsg {
    repeated AccelRGroupRingAddResponse response = 1;
}

message AccelRGroupRingDelRequest {
    string rgroup_name = 1;
    string ring_name = 2;
}

message AccelRGroupRingDelRequestMsg {
    repeated AccelRGroupRingDelRequest request = 1;
}

message AccelRGroupRingDelResponse {
    types.ApiStatus api_status = 1;
}

message AccelRGroupRingDelResponseMsg {
    repeated AccelRGroupRingDelResponse response = 1;
}

message AccelRGroupResetSetRequest {
    string rgroup_name = 1;
    uint32 sub_ring = 2;
    uint32 reset_sense = 3;
}
message AccelRGroupResetSetRequestMsg {
    repeated AccelRGroupResetSetRequest request = 1;
}

message AccelRGroupResetSetResponse {
    types.ApiStatus api_status = 1;
    uint32 last_ring_handle = 2;
    uint32 last_sub_ring = 3;
}

message AccelRGroupResetSetResponseMsg {
    repeated AccelRGroupResetSetResponse response = 1;
}

message AccelRGroupEnableSetRequest {
    string rgroup_name = 1;
    uint32 sub_ring = 2;
    uint32 enable_sense = 3;
}

message AccelRGroupEnableSetRequestMsg {
    repeated AccelRGroupEnableSetRequest request = 1;
}

message AccelRGroupEnableSetResponse {
    types.ApiStatus api_status = 1;
    uint32 last_ring_handle = 2;
    uint32 last_sub_ring = 3;
}

message AccelRGroupEnableSetResponseMsg {
    repeated AccelRGroupEnableSetResponse response = 1;
}

message AccelRGroupPndxSetRequest {
    string rgroup_name = 1;
    uint32 sub_ring = 2;
    uint32 val = 3;
    uint32 conditional = 4;
}

message AccelRGroupPndxSetRequestMsg {
    repeated AccelRGroupPndxSetRequest request = 1;
}

message AccelRGroupPndxSetResponse {
    types.ApiStatus api_status = 1;
    uint32 last_ring_handle = 2;
    uint32 last_sub_ring = 3;
}

message AccelRGroupPndxSetResponseMsg {
    repeated AccelRGroupPndxSetResponse response = 1;
}

message AccelRGroupInfoGetRequest {
    string rgroup_name = 1;
    uint32 sub_ring = 2;
}

message AccelRGroupInfoGetRequestMsg {
    repeated AccelRGroupInfoGetRequest request = 1;
}

message AccelRGroupRingInfoSpec {
    uint32 ring_handle = 1;
    uint32 sub_ring = 2;
    fixed64 base_pa = 3;
    fixed64 pndx_pa = 4;
    fixed64 shadow_pndx_pa = 5;
    fixed64 opaque_tag_pa = 6;
    uint32 opaque_tag_size = 7;
    uint32 ring_size = 8;
    uint32 desc_size = 9;
    uint32 pndx_size = 10;
    uint32 sw_reset_capable = 11;
    uint32 sw_enable_capable = 12;
}

message AccelRGroupInfoGetResponse {
    types.ApiStatus api_status = 1;
    repeated AccelRGroupRingInfoSpec ring_info_spec = 2;
}

message AccelRGroupInfoGetResponseMsg {
    repeated AccelRGroupInfoGetResponse response = 1;
}

message AccelRGroupIndicesGetRequest {
    string rgroup_name = 1;
    uint32 sub_ring = 2;
}

message AccelRGroupIndicesGetRequestMsg {
    repeated AccelRGroupIndicesGetRequest request = 1;
}

message AccelRGroupRingIndicesSpec {
    uint32 ring_handle = 1;
    uint32 sub_ring = 2;
    uint32 pndx = 3;
    uint32 cndx = 4;
    uint32 endx = 5;
}

message AccelRGroupIndicesGetResponse {
    types.ApiStatus api_status = 1;
    repeated AccelRGroupRingIndicesSpec ring_indices_spec = 2;
}

message AccelRGroupIndicesGetResponseMsg {
    repeated AccelRGroupIndicesGetResponse response = 1;
}

message AccelRGroupMetricsGetRequest {
    string rgroup_name = 1;
    uint32 sub_ring = 2;
}

message AccelRGroupMetricsGetRequestMsg {
    repeated AccelRGroupMetricsGetRequest request = 1;
}

message AccelRGroupRingMetricsSpec {
    uint32 ring_handle = 1;
    uint32 sub_ring = 2;
    uint64 input_bytes = 3;
    uint64 output_bytes = 4;
    uint64 soft_resets = 5;
}

message AccelRGroupMetricsGetResponse {
    types.ApiStatus api_status = 1;
    repeated AccelRGroupRingMetricsSpec ring_metrics_spec = 2;
}

message AccelRGroupMetricsGetResponseMsg {
    repeated AccelRGroupMetricsGetResponse response = 1;
}

message AccelRGroupMiscGetRequest {
    string rgroup_name = 1;
    uint32 sub_ring = 2;
}

message AccelRGroupMiscGetRequestMsg {
    repeated AccelRGroupMiscGetRequest request = 1;
}

message AccelRGroupMiscRegVal {
    string name = 1;
    uint32 val = 2;
}

message AccelRGroupRingMiscSpec {
    uint32 ring_handle = 1;
    uint32 sub_ring = 2;
    uint32 num_reg_vals = 3;
    repeated AccelRGroupMiscRegVal reg_val = 4;
}

message AccelRGroupMiscGetResponse {
    types.ApiStatus api_status = 1;
    repeated AccelRGroupRingMiscSpec ring_misc_spec = 2;
}

message AccelRGroupMiscGetResponseMsg {
    repeated AccelRGroupMiscGetResponse response = 1;
}

